README.md

```diff
@@ -1,55 +1,127 @@
-    $ source venv/bin/activate
-    ```
-    Once activated, you would see the shell prompt will change and it will show the name of the virtual environment that you are currently using. In our case it is venv.
+- ### Installing dependencies
 
-    Now install all the dependencies using:
-    ```
-    (venv) $ sudo pip3 install -r requirements.txt
-    ```
+  If you do not have `venv` package you can first install it:
 
-    Now run the flask server:
-    ```
-    (venv) $ python3 main.py
-    ```
+  ```
+  $ sudo apt install python3-venv
+  ```
+
+  Then to create new virtual environment run the following command in your project directory:
+
+  ```
+  $ sudo python3 -m venv venv
+  ```
+
+  You need to activate the virtual environment in order to use it. So activate the virtual environment using the following command:
+
+  ```
+  $ source venv/bin/activate
+  ```
 
+  Once activated, you would see the shell prompt will change and it will show the name of the virtual environment that you are currently using. In our case it is venv.
+
+  Now install all the dependencies using:
+
+  ```
+  (venv) $ sudo pip3 install -r requirements.txt
+  ```
+
+  Now run the flask server:
+
+  ```
+  (venv) $ python3 main.py
+  ```
 
 * ### Installing Google SDK
-    To install Google Cloud SDK for your OS: https://cloud.google.com/sdk/docs/install#installation_instructions
+
+  To install Google Cloud SDK for your OS: https://cloud.google.com/sdk/docs/install#installation_instructions
 
 * ### Deploying function to Google Cloud
-    Using the gcloud command-line tool, deploy your function from the directory containing your function code with the gcloud functions deploy command:
+
+  Using the gcloud command-line tool, deploy your function from the directory containing your function code with the gcloud functions deploy command:
+
+  ```
+  gcloud functions deploy FUNCTION_NAME --runtime python37 --trigger-http --allow-unauthenticated
+  ```
+
+  Note: <i> For more information please check - https://cloud.google.com/functions/docs/deploying/filesystem#python_example </i>
+
+## Testing
+
+- ### Test Python Google Cloud Functions on Your Local Development Environment
+
+  Using the Functions Framework for Python to debug and test how your HTTP-functions behave before deploying them to the Google Cloud Platform
+
+  1. Create a .env.local environment file in the Scripts directory
+  2. Define function name and a port number in this file
+
+  ```
+      export FUNCTION_NAME_HTTP=[name of the function to test]
+      export FUNCTION_PORT_HTTP=4000
+  ```
+
+  3. Create run-local-script.sh script in the scripts directory
+
+  ```
+  #! /bin/bash
+
+  DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+  SOURCE_FILE="${DIR}/../src/main.py"
+
+  source "${DIR}/.env.local"
+
+  functions-framework \
+      --source=${SOURCE_FILE} \
+      --target=${FUNCTION_NAME_HTTP} \
+      --signature-type=http \
+      --port=${FUNCTION_PORT_HTTP} \
+      --debug
+  ```
```

Comment:
This is super cool! You should make a separate bash script and check it in.

Reply:
I have the script but in .gitignore. But I think students can just copy paste it and get it running. I tried to keep instructions simple.